test/lib/jdk/test/lib/cds/CDSAppTester.java

```diff
@@ -0,0 +1,233 @@
+    abstract public String[] appCommandLine(RunMode runMode);
+
+    // optional
+    public void checkExecution(OutputAnalyzer out, RunMode runMode) throws Exception {}
+
+    private Workflow workflow;
+
+    public final boolean isStaticWorkflow() {
+        return workflow == Workflow.STATIC;
+    }
+
+    public final boolean isDynamicWorkflow() {
+        return workflow == Workflow.DYNAMIC;
+    }
+
+    private String classListLog() {
+        return "-Xlog:class+load=debug:file=" + classListFile + ".log";
+    }
+    private String staticDumpLog() {
+        return "-Xlog:cds=debug,cds+class=debug,cds+heap=warning,cds+resolve=debug:file=" + staticArchiveFile + ".log::filesize=0";
+    }
+    private String dynamicDumpLog() {
+        return "-Xlog:cds=debug,cds+class=debug,cds+resolve=debug,class+load=debug:file=" + dynamicArchiveFile + ".log::filesize=0";
+    }
+
+    private final String name;
+    private final String classListFile;
+    private final String staticArchiveFile;
+    private final String dynamicArchiveFile;
+
+    private String productionRunLog() {
+        return "-Xlog:cds:file=" + name() + ".production.log::filesize=0";
+    }
+
+    public CDSAppTester(String name) {
+        // Old workflow
+        this.name = name;
+        classListFile = name() + ".classlist";
+        staticArchiveFile = name() + ".static.jsa";
+        dynamicArchiveFile = name() + ".dynamic.jsa";
+    }
+
+    private void listOutputFile(String file) {
+        File f = new File(file);
+        if (f.exists()) {
+            System.out.println("[output file: " + file + " " + f.length() + " bytes]");
+        } else {
+            System.out.println("[output file: " + file + " does not exist]");
+        }
+    }
+
+    private void checkExecutionHelper(OutputAnalyzer output, RunMode runMode) throws Exception {
+        output.shouldHaveExitValue(0);
+        CDSTestUtils.checkCommonExecExceptions(output);
+        checkExecution(output, runMode);
+    }
+
+    private OutputAnalyzer createClassList() throws Exception {
+        RunMode runMode = RunMode.CLASSLIST;
+        String[] cmdLine = StringArrayUtils.concat(vmArgs(runMode), classListLog(),
+                                                   "-Xshare:off",
+                                                   "-XX:DumpLoadedClassList=" + classListFile,
+                                                   "-cp", classpath(runMode));
+        cmdLine = StringArrayUtils.concat(cmdLine, appCommandLine(runMode));
+
+        ProcessBuilder pb = ProcessTools.createTestJavaProcessBuilder(cmdLine);
+        Process process = pb.start();
+        OutputAnalyzer output = CDSTestUtils.executeAndLog(process, "classlist");
+        listOutputFile(classListFile);
+        listOutputFile(classListFile + ".log");
+        checkExecutionHelper(output, runMode);
+        return output;
+    }
+
+    private OutputAnalyzer dumpStaticArchive() throws Exception {
```

Comment:
The code from 156 to 162 is repeated 3 times here, is it worth making another function for this?